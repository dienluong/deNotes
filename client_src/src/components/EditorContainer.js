import { connect } from 'react-redux';
import Editor from './widgets/Editor';
import { changeContentAction } from '../redux/actions/editorActions';

function mapStateToProps(state) {
  return {
    id: state.editorContent.id,
    delta: state.editorContent.delta,
    content: state.editorContent.content,
    dateCreated: state.editorContent.dateCreated,
    dateModified: state.editorContent.dateModified,
  };
}

function mapDispatchToProps(dispatch) {
  return {
    contentChangeHandler(content, delta, source, editor) {
      return dispatch(changeContentAction({ editor, content }));
    },
  };
}

// TODO: remove
// function mergeProps(stateProps, dispatchProps, ownProps) {
//   return Object.assign({}, ownProps, stateProps, dispatchProps);
// }

// Note: ownProps is automatically forwarded to the wrapped component
const EditorContainer = connect(mapStateToProps, mapDispatchToProps)(Editor);
export default EditorContainer;
