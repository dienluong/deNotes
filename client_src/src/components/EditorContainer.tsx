import { connect } from 'react-redux';
import Editor from './widgets/Editor';
import { changeContentAction } from '../redux/actions/editorActions';

// Types
import { ThunkDispatch } from 'redux-thunk';
import { AnyAction } from 'redux';
import { Delta, Sources, BoundsStatic, RangeStatic, DeltaStatic } from 'quill';
interface UnprivilegedEditor {
  getLength(): number;
  getText(index?: number, length?: number): string;
  getHTML(): string;
  getBounds(index: number, length?: number): BoundsStatic;
  getSelection(focus?: boolean): RangeStatic;
  getContents(index?: number, length?: number): DeltaStatic;
}
interface DispatchProps {
  contentChangeHandler: (content: string, delta: Delta, source: Sources, editor: UnprivilegedEditor) => AnyAction;
}

function mapStateToProps(state: AppStateT) {
  return {
    id: state.editorContent.id,
    title: state.editorContent.title,
    delta: state.editorContent.delta,
    content: state.editorContent.content,
    dateCreated: state.editorContent.dateCreated,
    dateModified: state.editorContent.dateModified,
    readOnly: state.editorContent.readOnly,
  };
}

function mapDispatchToProps(dispatch: ThunkDispatch<AppStateT, any, AnyAction> ): DispatchProps {
  return {
    contentChangeHandler(content, delta, source, editor) {
      return dispatch(changeContentAction({ editor, content }));
    },
  };
}

// Note: ownProps is automatically forwarded to the wrapped component
const EditorContainer = connect(mapStateToProps, mapDispatchToProps)(Editor);
export default EditorContainer;
