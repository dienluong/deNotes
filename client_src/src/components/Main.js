import React from 'react';
// import { BrowserRouter as Router, Route } from 'react-router-dom';
import Editor from './Editor';
import NotesList from './NotesList';
import './Main.css';

class Main extends React.Component {
  constructor(props) {
    super(props);
  }

  render() {
    return (
      <div>
        <NotesList />
        <Editor />
      </div>
    );
  }
}

export default Main;
