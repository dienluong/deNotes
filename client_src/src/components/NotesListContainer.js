import { connect } from 'react-redux';
import NotesList from './widgets/NotesList';
import {
  selectNodeAction,
  navigatePathAction,
  changeNotesTreeAction,
  changeNodeTitleAction,
  deleteNodeAction,
  addNoteAction,
  switchActiveNodeOnDelete,
  switchActiveNodeOnAdd,
  addAndSelectNode,
} from '../redux/actions/notesListActions';

import { getNodeKey, translatePathToInfo } from '../utils/treeUtils';

function mapStateToProps(state) {
  const activePath = translatePathToInfo({ notesTree: state.notesTree, path: state.activeNode.path, kind: 'title' });
  return {
    notesTree: state.notesTree,
    activeNode: state.activeNode,
    activePath,
  };
}

function mapDispatchToProps(dispatch) {
  function toolbarNewFolderBtnClickHandler() {
    return dispatch(addAndSelectNode({ kind: 'folder' }));
  }

  function toolbarNewNoteBtnClickHandler() {
    return dispatch(addAndSelectNode({ kind: 'item' }));
  }

  const toolbarHandlersMap = new Map();

  toolbarHandlersMap.set('New Folder', toolbarNewFolderBtnClickHandler);
  toolbarHandlersMap.set('New Note', toolbarNewNoteBtnClickHandler);

  return {
    treeChangeHandler(notesTree) {
      return dispatch(changeNotesTreeAction(notesTree));
    },
    nodeTitleChangeHandler({ node, title, path }) {
      return dispatch(changeNodeTitleAction({ node, title, path }));
    },
    pathNavigatorClickHandler({ idx }) {
      return dispatch(navigatePathAction({ idx }));
    },
    nodeClickHandler({ id = null, path = [] }) {
      return dispatch(selectNodeAction({ id, path }));
    },
    deleteNodeBtnHandler({ node, path }) {
      dispatch(deleteNodeAction({ node, path }));
      return dispatch(switchActiveNodeOnDelete({ id: node.id, path }));
    },
    addNoteBtnHandler({ path }) {
      dispatch(addNoteAction({ path }));
      return dispatch(switchActiveNodeOnAdd({ path }));
    },
    toolbarHandlersMap: toolbarHandlersMap,
  };
}

function mergeProps(stateProps, dispatchProps, ownProps) {
  return Object.assign({}, ownProps, stateProps, dispatchProps, { getNodeKey });
}

const NotesListContainer = connect(mapStateToProps, mapDispatchToProps, mergeProps)(NotesList);
export default NotesListContainer;
