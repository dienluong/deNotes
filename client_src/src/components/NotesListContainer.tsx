import { connect } from 'react-redux';
import { selectTitlesFromActivePath, selectSiblingsOfActiveNode } from '../redux/selectors';
import * as rootReducer from '../redux/reducers';
import NotesList from './widgets/NotesList';
import {
  selectNodeThunkAction,
  navigatePathThunkAction,
  changeNotesFolderThunkAction,
  changeNodeTitleAction,
  deleteNodeThunkAction,
  addAndSelectNodeThunkAction,
} from '../redux/actions/notesListActions';

import { getNodeKey } from '../utils/treeUtils';

// Types
import { ThunkDispatch } from 'redux-thunk';
import { AnyAction } from 'redux';
interface DispatchProps {
  treeChangeHandler: (params: any) => AnyAction;
  nodeTitleChangeHandler: (params: { node: TreeNodeT, title: string, path: TreeNodePathT }) => AnyAction;
  pathNavigatorClickHandler: (params: { idx: number }) => AnyAction;
  nodeClickHandler: (params: any) => AnyAction;
  deleteNodeBtnHandler: (params: any) => Promise<AnyAction>;
  addNoteBtnHandler: (params: any) => AnyAction;
  toolbarHandlersMap: Map<string, () => AnyAction>;
}

export const TOOLBAR_LABELS = {
  NEW_FOLDER: 'New Folder',
  NEW_NOTE: 'New Note',
};

function mapStateToProps(state: AppStateT) {
  const activePathByTitles = selectTitlesFromActivePath(state);

  // const activePath = translatePathToInfo({ notesTree: state.notesTree, path: state.activeNode.path, kind: 'title' });
  if (mapStateToProps.cache.notesTree !== rootReducer.selectNotesTree(state)) {
    console.log('~~~~~~~~~~~~~~~~ Notes Tree Changed');
    mapStateToProps.cache.notesTree = rootReducer.selectNotesTree(state);
  } else {
    console.log('~~~~~~~~~~~~~~~~ Same Notes Tree');
  }
  if (mapStateToProps.cache.activeNode !== rootReducer.selectActiveNode(state)) {
    console.log('~~~~~~~~~~~~~~~~ Active Node Changed');
    mapStateToProps.cache.activeNode = rootReducer.selectActiveNode(state);
  } else {
    console.log('~~~~~~~~~~~~~~~~ Same Active Node');
  }
  if (mapStateToProps.cache.activePath !== activePathByTitles) {
    console.log('~~~~~~~~~~~~~~~~ Active Path Changed');
    mapStateToProps.cache.activePath = activePathByTitles;
  } else {
    console.log('~~~~~~~~~~~~~~~~ Same Active Path');
  }
  if (mapStateToProps.cache.editorContent !== rootReducer.selectEditorContent(state)) {
    console.log('~~~~~~~~~~~~~~~~ Editor Content Changed');
    mapStateToProps.cache.editorContent = rootReducer.selectEditorContent(state);
  } else {
    console.log('~~~~~~~~~~~~~~~~ Same Editor Content');
  }
  // TODO: Remove ABOVE

  // Get the siblings of the current active node (including itself). The children will be passed as prop to the component
  const children = selectSiblingsOfActiveNode(state);
  return {
    tree: children,
    activeNode: rootReducer.selectActiveNode(state),
    activePath: activePathByTitles,
  };
}

// TODO: Remove
// Memoization.
mapStateToProps.cache = {} as { notesTree: any, activeNode: any, activePath: any, editorContent: any };

function mapDispatchToProps(dispatch: ThunkDispatch<AppStateT, any, AnyAction>): DispatchProps {
  function toolbarNewFolderBtnHandler() {
    return dispatch(addAndSelectNodeThunkAction({ kind: 'folder' }));
  }

  function toolbarNewNoteBtnHandler() {
    return dispatch(addAndSelectNodeThunkAction({ kind: 'item' }));
  }

  const toolbarHandlersMap = new Map();

  toolbarHandlersMap.set(TOOLBAR_LABELS.NEW_FOLDER, toolbarNewFolderBtnHandler);
  toolbarHandlersMap.set(TOOLBAR_LABELS.NEW_NOTE, toolbarNewNoteBtnHandler);

  return {
    treeChangeHandler(tree) {
      return dispatch(changeNotesFolderThunkAction({ folder: tree }));
    },
    nodeTitleChangeHandler({ node, title, path }) {
      // We are not using the path received from the NotesList component because that path is not for the entire tree;
      // remember that NotesList only receives and renders a given leaf of the whole tree.
      return dispatch(changeNodeTitleAction({ node, title }));
    },
    pathNavigatorClickHandler({ idx }) {
      return dispatch(navigatePathThunkAction({ idx }));
    },
    nodeClickHandler({ id = '', path = [] }) {
      // We are not using the path received from the NotesList component because that path is not for the entire tree;
      // remember that NotesList only receives and renders a given leaf of the whole tree.
      return dispatch(selectNodeThunkAction({ id }));
    },
    deleteNodeBtnHandler({ node, path }) {
      // We are not using the path received from the NotesList component because that path is not for the entire tree;
      // remember that NotesList only receives and renders a given leaf of the whole tree.
      return dispatch(deleteNodeThunkAction({ node }));
    },
    addNoteBtnHandler({ path } ) {
      // We are not using the path received from the NotesList component because that path is not for the entire tree;
      // remember that NotesList only receives and renders a given leaf of the whole tree.
      return dispatch(addAndSelectNodeThunkAction({ kind: 'item' }));
    },
    toolbarHandlersMap: toolbarHandlersMap,
  };
}

function mergeProps(stateProps: object, dispatchProps: object, ownProps: object) {
  return Object.assign({}, ownProps, stateProps, dispatchProps, { getNodeKey });
}

const NotesListContainer = connect(mapStateToProps, mapDispatchToProps, mergeProps)(NotesList);
export default NotesListContainer;
