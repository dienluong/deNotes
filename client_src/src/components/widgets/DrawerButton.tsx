import React from 'react';
import Fab from '@material-ui/core/Fab';
import MenuIcon from '@material-ui/icons/Menu';
import Zoom from '@material-ui/core/Zoom';

function DrawerButton({ visible, clickHandler, className }: { visible: boolean, clickHandler: () => unknown, className: string }) {
  return (
    <div className={ className }>
      <Zoom in={ visible } unmountOnExit>
        <Fab aria-label="notes list drawer" onClick={ clickHandler }>
          <MenuIcon />
        </Fab>
      </Zoom>
    </div>
  );
}

export default DrawerButton;
