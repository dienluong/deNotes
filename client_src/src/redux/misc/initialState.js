import uuid from 'uuid/v1';
import Delta from 'quill-delta';
const _ID_DELIMITER = process.env.REACT_APP_ID_DELIMITER;

// TODO: Is it the best place to define this?
const _rootNode = [
  {
    title: '/',
    subtitle: '',
    uniqid: uuid(),
    get id() {
      return `${this.type}${_ID_DELIMITER}${this.uniqid}`;
    },
    type: 'folder',
    expanded: true,
    children: [],
  },
];

const epoch = 0;
export default {
  notesTree: _rootNode,
  activeNode: {
    id: _rootNode[0].id,
    path: [_rootNode[0].id],
  },
  editorContent: {
    id: '',
    title: '',
    content: '<p><br></p>',
    delta: new Delta(),
    dateCreated: epoch,
    dateModified: epoch,
    readOnly: false,
  },
};
