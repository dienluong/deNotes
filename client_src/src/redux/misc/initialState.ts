import uuid from 'uuid/v4';
import Delta from 'quill-delta';
import { nodeTypes, DEFAULT_ID_DELIMITER} from '../../utils/appCONSTANTS';

const _ID_DELIMITER = process.env.REACT_APP_ID_DELIMITER || DEFAULT_ID_DELIMITER;

const _rootNode: TreeNodeT = {
    title: '/',
    subtitle: '',
    uniqid: uuid(),
    get id() {
      return `${this.type}${_ID_DELIMITER}${this.uniqid}`;
    },
    type: nodeTypes.FOLDER,
    expanded: true,
    children: [],
};

const epoch = 0;
const now = Date.now();
const initialState: AppStateT = {
  userInfo: {
    id: 'default-user',
  },
  notesTree: {
    id: uuid(),
    tree: [_rootNode],
    editMode: false,
    editModeSelectedNodes: [],
    dateCreated: now,
    dateModified: now,
  },
  // activeNode contains the ID of the current active node, and its path, in the tree;
  // the active node can be an item (note) or a folder;
  // if the active node is the root (with no child selected), then ID is constant NONE_SELECTED and path is [NONE_SELECTED]
  activeNode: {
    id: _rootNode.id,
    path: [_rootNode.id],
  },
  editorContent: {
    id: 'default',
    title: '',
    content: '<p><br></p>',
    delta: new Delta(),
    dateCreated: epoch,
    dateModified: epoch,
    readOnly: true,
  },
  modalInfo: {
    type: '',
    props: {},
  }
};

export default initialState;
