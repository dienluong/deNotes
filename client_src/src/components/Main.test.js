import React from 'react';
import Main from './Main';
import EditorContainer from './EditorContainer';
import NotesListContainer from './NotesListContainer';
import DrawerButton from './widgets/DrawerButton';
import Fab from '@material-ui/core/Fab';
jest.mock('@material-ui/core/useMediaQuery');
import { unstable_useMediaQuery as useMediaQuery } from '@material-ui/core/useMediaQuery';
import createMockStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import initialState from '../redux/misc/initialState';
import { createSerializer } from 'enzyme-to-json';

// Configure Enzyme
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
configure({ adapter: new Adapter() });
let wrapper = {};

// Configure snapshot serializer
expect.addSnapshotSerializer(createSerializer());

beforeAll(() => {
  useMediaQuery.mockImplementation(() => {
    return false;
  });
});

afterEach(() => {
  if (typeof wrapper.unmount === 'function') {
    wrapper.unmount();
  }
});

it('renders Main correctly using EditorContainer, NotesListContainer and DrawerButton', () => {
  let store = createMockStore([thunk])(initialState);
  wrapper = shallow(<Provider store={ store }><Main /></Provider>).dive({ context: { store } });
  expect(wrapper).toMatchSnapshot();
  expect(wrapper.find(EditorContainer).exists()).toBe(true);
  expect(wrapper.find(NotesListContainer).exists()).toBe(true);
  expect(wrapper.find(DrawerButton).exists()).toBe(true);
});

it('sets size to medium and disables minimalist style for width larger than 600px', () => {
  useMediaQuery.mockImplementation(query => {
    return !(query === '(max-width:600px)');
  });
  let store = createMockStore([thunk])(initialState);
  wrapper = shallow(<Provider store={ store }><Main /></Provider>).dive({ context: { store } });
  expect(wrapper.find(EditorContainer).prop('minimalist')).toBe(false);
  expect(wrapper.find(NotesListContainer).prop('size')).toBe('medium');
  expect(wrapper.find(DrawerButton).prop('size')).toBe('medium');
});

it('sets size to small and enables minimalist style when max width is 600px', () => {
  useMediaQuery.mockImplementation(query => {
    return query === '(max-width:600px)';
  });
  let store = createMockStore([thunk])(initialState);
  wrapper = shallow(<Provider store={ store }><Main /></Provider>).dive({ context: { store } });
  expect(wrapper.find(EditorContainer).prop('minimalist')).toBe(true);
  expect(wrapper.find(NotesListContainer).prop('size')).toBe('small');
  expect(wrapper.find(DrawerButton).prop('size')).toBe('small');
});

it('opens drawer when drawer button clicked', () => {
  let store = createMockStore([thunk])(initialState);
  wrapper = shallow(<Provider store={ store }><Main /></Provider>).dive({ context: { store } });
  expect(wrapper.find(NotesListContainer).prop('drawerOpen')).toBe(false);
  expect(wrapper.find(DrawerButton).prop('visible')).toBe(true);
  wrapper.find(DrawerButton).dive().find(Fab).simulate('click');
  expect(wrapper.find(NotesListContainer).prop('drawerOpen')).toBe(true);
  expect(wrapper.find(DrawerButton).prop('visible')).toBe(false);
});
