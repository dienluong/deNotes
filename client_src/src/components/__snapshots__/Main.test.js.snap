// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Main correctly using EditorContainer, NotesListContainer and DrawerButtonContainer 1`] = `
<div
  className="dnt__main-root"
>
  <main
    className="dnt__main-editor"
  >
    <WithStyles(Paper)
      elevation={1}
    >
      <Connect(Editor)
        minimalist={false}
        options={
          Object {
            "placeholder": "Welcome to deNotes! v0.02",
          }
        }
      />
    </WithStyles(Paper)>
  </main>
  <nav
    className="dnt__main-nav"
  >
    <Connect(NotesListDrawer)
      drawerSide="left"
      size="medium"
    />
  </nav>
  <Connect(DrawerButton)
    className="dnt__main-nav-btn"
    size="medium"
  />
</div>
`;
